src/sknnr/transformers/_cca.py

```diff
@@ -13,228 +22,85 @@ def zero_sum_columns(arr: NDArray) -> NDArray:
-        return np.linalg.lstsq(R, right, rcond=None)
-
-    @property
-    def Y_fit(self):
-        return self.X_scale @ self._least_squares[0]
-
-    @property
-    def sol(self):
-        return np.linalg.svd(self.Y_fit, full_matrices=False)
-
-    @property
-    def rank(self):
-        ls_rank = self._least_squares[2]
-        pos_s = np.sum([self.sol[1] > self.ZERO])
-        return min(ls_rank, pos_s)
-
-    @property
-    def _U(self):
-        return self.sol[0][:, : self.rank]
-
-    @property
-    def S(self):
-        return np.square(self.sol[1])[: self.rank]
-
-    @property
-    def _V(self):
-        return self.sol[2].T[:, : self.rank]
-
-    @property
-    def U(self):
-        return self._U / np.sqrt(self.rw)[:, None]
+    def __init__(self, X, Y):
+        self.X = np.array(X)
+        self.Y = np.array(Y)
+        self._check_inputs()
+        self._transform_Y()
+
+    @abstractmethod
+    def _transform_Y(self):
+        ...
+
+    @abstractmethod
+    def _transform_X(self):
+        ...
+
+    def _check_inputs(self):
+        if not is_2d_numeric_numpy_array(self.X):
+            msg = "X must be a 2D numeric numpy array"
+            raise ValueError(msg)
 
-    @property
-    def V(self):
-        return self._V / np.sqrt(self.cw)[:, None]
+        if not is_2d_numeric_numpy_array(self.Y):
+            msg = "Y must be a 2D numeric numpy array"
+            raise ValueError(msg)
 
-    @property
-    def _WA(self):
-        x = np.diag(1.0 / np.sqrt(self.S))
-        return self.transform.Y.dot(self._V).dot(x)
+        if self.X.shape[0] != self.Y.shape[0]:
+            msg = "X and Y must have the same number of rows"
+            raise ValueError(msg)
 
-    @property
-    def WA(self):
-        return self._WA / np.sqrt(self.rw)[:, None]
+        if np.any(zero_sum_rows(self.X)):
+            msg = "All row sums in X must be greater than 0"
+            raise ValueError(msg)
 
-    @property
-    def biplot_scores(self):
-        a = 1.0 / np.sqrt(np.square(self.X_scale).sum(axis=0))
-        b = self.X_scale.T.dot(self._U)
-        return a[:, None] * b
+        excluded_columns = zero_sum_columns(self.Y)
+        self.Y = self.Y[:, ~excluded_columns]
 
-    @property
-    def Y_resid(self):
-        return self.transform.Y - self.Y_fit
+    def __call__(self):
+        X_scale = self._transform_X()
+        Q, R = np.linalg.qr(X_scale)
+        solution, _, rank, _ = np.linalg.lstsq(R, Q.T @ self.Y, rcond=None)
+        Y_fit = X_scale @ solution
+        U, S, _ = np.linalg.svd(Y_fit, full_matrices=False)
+        self.rank = min(rank, np.sum([S > self.ZERO]))
+        self.eigenvalues = np.square(S)[: self.rank]
+        self.axis_weights = np.diag(np.sqrt(self.eigenvalues / self.eigenvalues.sum()))
+        self.coefficients = np.linalg.lstsq(R, Q.T @ U[:, : self.rank], rcond=None)[0]
+        return self
 
     @property
     def max_components(self):
         return self.rank
```

Comment:
What do you think about replacing `self.rank` with `self.max_components` to avoid needing this property as an "alias"? It would make `__call__` a little more wordy.

Reply:
I hemmed and hawed on this one for a while.  I don't think the latter suggestion works because `CCorA` doesn't return `rank` as its `max_components` but rather the number of "significant" components.  But the former suggestion would definitely work.  I haven't done so and my logic might be a bit shaky, so see what you think.

I've kept the term `rank` in there to be as consistent as possible with the variable names in `vegan` and because `rank` would be the default name when running `np.linalg.lstsq` (even though we actually modify it later on for positive eigenvalues).  I felt like the `max_components` property was lightweight and inobtrusive enough to have it alias `rank` but not completely replace it.  Clean-code-wise, it definitely makes sense to do as you say, but having `rank` in there might be a valuable reminder for me.